allennlp/modules/elmo.py

```diff
@@ -515,3 +565,63 @@ def forward(self,  # pylint: disable=arguments-differ
                 'activations': output_tensors,
                 'mask': mask,
         }
+
+    def create_cached_cnn_embeddings(self, tokens: List[str]) -> None:
+        """
+        Given a list of tokens, this method precomputes word representations
+        by running just the character convolutions and highway layers of elmo,
+        essentially creating uncontextual word vectors. On subsequent forward passes,
+        if all of the words in the batch have cached, the word ids are looked up from
```

Comment:
I didn't see anything that actually checked "if all of the words in the batch have [been] cached".  Did I miss something, or is this comment wrong?

Reply:
This was from an old version where you could "back off" to computing them on the fly. Actually I think I need to add this, because otherwise you can't deal with new words, e.g at test time.

Reply:
Ugh this completely breaks this approach actually, because it requires access to the character ids if the word ids aren't in the cache, which means to back off to a different approach, both need to be passed. That's pretty frustrating. This is basically useless if it can't handle new words at inference time, because it defeats a big bonus of using elmo.

Reply:
Could you have this take the whole `TextField` input, and use whichever part you need?  You could also imagine dynamically growing the cache.  But that's probably not necessary, because the words that aren't in your cache are always going to be OOV, so you can just filter them that way...

Reply:
I don't see how that's an option? We need to be able to use it as a token embedder, right?

Reply:
Well, there's no reason we couldn't just add an `ElmoTextFieldEmbedder` that handled things differently from the `BasicTextFieldEmbedder`, and passed multiple representations on to the elmo classes.  And we say that you need to do this if you want caching.  Things get a bit complicated, so if you're not too worried about speed, you can just do things as they currently are, but if you want to speed things up with caching, you modify your config even more and use this other `TextFieldEmbedder`.

These are just some thoughts that I'm hoping will help you figure out a good solution; I'm not sure whether this is the right way to do things or not.

Reply:
Sorry, I didn't mean to be obnoxious - it's just i've tried to do this twice now and both times run into kind of complicated issues which was annoying me.

That way of doing things seems plausible, I think. I'll try that today.